src/listener.js

```diff
@@ -0,0 +1,107 @@
+'use strict'
+
+const http = require('http')
+const EventEmitter = require('events')
+const log = require('debug')('libp2p:libp2p:webrtcdirect:listener')
+
+const isNode = require('detect-node')
+const wrtc = require('wrtc')
+const SimplePeer = require('simple-peer')
+
+const multibase = require('multibase')
+
+const toConnection = require('./socket-to-conn')
+
+module.exports = ({ handler, upgrader }, options = {}) => {
+  const listener = new EventEmitter()
+  const server = http.createServer()
+
+  let maSelf
+
+  // Keep track of open connections to destroy in case of timeout
+  server.__connections = []
+
+  server.on('request', async (req, res) => {
+    res.setHeader('Content-Type', 'text/plain')
+    res.setHeader('Access-Control-Allow-Origin', '*')
+
+    const path = req.url
+    const incSignalStr = path.split('?signal=')[1]
+    const incSignalBuf = multibase.decode(Buffer.from(incSignalStr))
+    const incSignal = JSON.parse(incSignalBuf.toString())
+
+    options = {
+      ...options,
+      trickle: false
+    }
+
+    if (isNode) {
+      options.wrtc = wrtc
+    }
+
+    const channel = new SimplePeer(options)
+    const maConn = toConnection(channel)
+    log('new inbound connection %s', maConn.remoteAddr)
+
+    const conn = await upgrader.upgradeInbound(maConn)
+    log('inbound connection %s upgraded', maConn.remoteAddr)
+
+    trackConn(server, maConn)
```

Comment:
We're not untracking connections when they close, we should add a test for this.